src/impl/tpls/KokkosSparse_spmv_tpl_spec_decl.hpp

```diff
@@ -44,9 +44,214 @@
+					(const float *) A.values.data(), A.graph.row_map.data(), A.graph.entries.data(),
+					(const float *) x.data(),
+					(const float *) &beta,
+					(float *) y.data());
+
+      } else  if (std::is_same<value_type,double>::value) {
+	cusparseStatus = cusparseDcsrmv(cusparseHandle, CUSPARSE_OPERATION_NON_TRANSPOSE,
+					A.numRows(), A.numCols(), A.nnz(),
+					(double const *) &alpha, descrA,
+					(double const *) A.values.data(), A.graph.row_map.data(), A.graph.entries.data(),
+					(double const *) x.data(),
+					(double const *) &beta,
+					(double *) y.data());
+      } else {
+	throw std::logic_error("Trying to call cusparse SpMV with a scalar type that is not float or double!");
+      }
+    } else {
+      throw std::logic_error("Trying to call cusparse SpMV with an offset type that is not int!");
+    }
+
+    cusparseStatus = cusparseDestroyMatDescr(descrA);
+    if (cusparseStatus != CUSPARSE_STATUS_SUCCESS) {
+      throw("KokkosSparse::spmv[TPL_CUSPARSE,double]: matrix descriptor was not desctroyed correctly");
+    }
+    cusparseStatus = cusparseDestroy(cusparseHandle);
+    cusparseHandle = 0;
+    if (cusparseStatus != CUSPARSE_STATUS_SUCCESS) {
+      throw("KokkosSparse::spmv[TPL_CUSPARSE,double]: cusparse handle was not desctroyed correctly");
+    }
+
+#endif // CUSPARSE_VERSION
+  }
+
+#define KOKKOSSPARSE_SPMV_CUSPARSE(SCALAR, OFFSET, LAYOUT, COMPILE_LIBRARY) \
+  template<>								\
+  struct SPMV<SCALAR const,  OFFSET const, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<Kokkos::Unmanaged>, OFFSET const, \
+	      SCALAR const*, LAYOUT,       Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<Kokkos::Unmanaged|Kokkos::RandomAccess>, \
+	      SCALAR*,       LAYOUT,       Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<Kokkos::Unmanaged>, \
+	      true, COMPILE_LIBRARY> {					\
+    using device_type = Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>; \
+    using memory_trait_type = Kokkos::MemoryTraits<Kokkos::Unmanaged>;	\
+    using AMatrix = CrsMatrix<SCALAR const, OFFSET const, device_type, memory_trait_type, OFFSET const>; \
+    using XVector = Kokkos::View<SCALAR const*, LAYOUT,device_type, Kokkos::MemoryTraits<Kokkos::Unmanaged|Kokkos::RandomAccess>>; \
+    using YVector = Kokkos::View<SCALAR*, LAYOUT, device_type, memory_trait_type>; \
+									\
+    using coefficient_type = typename YVector::non_const_value_type;	\
+									\
+    static void spmv (const char mode[],				\
+		      const coefficient_type& alpha,			\
+		      const AMatrix& A,					\
+		      const XVector& x,					\
+		      const coefficient_type& beta,			\
+		      const YVector& y) {				\
+      Kokkos::Profiling::pushRegion("KokkosSparse::spmv[TPL_CUSPARSE,double]");	\
+      spmv_cusparse(mode, alpha, A, x, beta, y);			\
+      Kokkos::Profiling::popRegion();					\
+    }									\
+  };
+
+  KOKKOSSPARSE_SPMV_CUSPARSE(double, int, Kokkos::LayoutLeft,  true)
```

Comment:
Is complex support to be added later?

Reply:
Potentially but I am a bit worried about the alignment of `cuComplex` and that of `Kokkos::complex` for that reason I would like to defer this to allow careful testing when that feature goes in.

Reply:
I'll need to look a bit more carefully at the ETI TPL layer you added, but if complex is not enabled then does this mean kokkos-kernels spmv will use the existing fallback implementation? If not, that'll need to be addressed or else this will cause issues in Trilinos if anyone enables cusparse.

Reply:
If you look at `KokkosSparse_spmv_tpl_spec_avail.hpp` you will see that the tpl is not made available for complex type so it will fall back to kokkos-kernel implementation of spmv in that case.

Reply:
@ndellingwood Let me second Luc's concern about alignment of complex.  I remember y'all added an option to control alignment of `Kokkos::complex`; we could use that to decide whether to enable complex support.